### PR TITLE
Fix test bugs

### PR DESCRIPTION
**Pull request by jessberg**
These are miscellaneous changes I thought would be good to make while I was making end to end tests.  They are:
1. I made the main graph look more like the hipster shop demo;  it's a little more realistic now.
2. I added some more explanatory error messages and extra error checking when adding edges.
3. I fixed a bug wherein the plugin wasn't connected to its node.
4. I changed the plugin's stored node to be a vector of nodes;  now that the filter can output more than one RPC, it follows that the plugin should be able to store more than one RPC.

### CODE REVIEW

**Comment by fruffy** [comment]
This needs to be rebased.

File: sim/src/main.rs [hunk 1]
```diff
@@ -50,40 +50,73 @@ fn main() {
...
+        "productcatalogservice-v1",
+        "currencyservice-v1",
+        "paymentservice-v1",
+        "shippingservice-v1",
+        "emailservice-v1",
+        "checkoutservice-v1",
+        "recomendationservice-v1",
+        "adservice-v1",
+    ]
+    .to_vec();
```
Comment (fruffy) [inline sim/src/main.rs:64]: Why are we using hipster shop instead of bookinfo? Just for this example? In tracing_env we should use the bookinfo though.

File: sim/src/main.rs [hunk 2]
```diff
@@ -50,40 +50,73 @@ fn main() {
...
+        // node: ID, capacity, egress rate, generation rate, plugin
+        simulator.add_node(node, 10, 5, 0, plugin_str);
+    }
+    simulator.add_node("loadgenerator-v1", 10, 1, 1, None);
+    simulator.add_node("sink", 10, 1, 0, None); // rpc sink
+    simulator.add_storage(storage_name);
+
+    // add all connections to storage and to the sink - we want traces to be able to end arbitrarily
+    for node in &regular_nodes {
+        simulator.add_edge(1, node, storage_name, true);
```
Comment (fruffy) [inline sim/src/main.rs:78]: Does this imply requests can also be sent to storage? This may lead to strange results.

File: sim/src/main.rs [hunk 3]
```diff
@@ -50,40 +50,73 @@ fn main() {
...
+    .to_vec();
+    for node in &regular_nodes {
+        // node: ID, capacity, egress rate, generation rate, plugin
+        simulator.add_node(node, 10, 5, 0, plugin_str);
+    }
+    simulator.add_node("loadgenerator-v1", 10, 1, 1, None);
+    simulator.add_node("sink", 10, 1, 0, None); // rpc sink
+    simulator.add_storage(storage_name);
+
+    // add all connections to storage and to the sink - we want traces to be able to end arbitrarily
```
Comment (fruffy) [inline sim/src/main.rs]: How can a trace end arbitrarily?

File: sim/src/node.rs [hunk 4]
```diff
@@ -76,12 +76,15 @@ impl SimElement for Node {
...
-                        ret.push((rpc, dest));
+                        ret.push((rpc, dest.clone()));
                         break;
                     }
                 }
                 if !have_dest {
-                    print!("WARNING:  RPC given with invalid destination\n");
+                    print!(
+                        "WARNING:  RPC given with invalid destination {0}\n",
+                        dest.clone()
```
Comment (fruffy) [inline sim/src/node.rs]: Cloning for a print seems wrong.

File: sim/src/node.rs [hunk 5]
```diff
@@ -76,12 +76,15 @@ impl SimElement for Node {
                 for n in &self.neighbors {
                     if n == &dest {
                         have_dest = true;
-                        ret.push((rpc, dest));
+                        ret.push((rpc, dest.clone()));
```
Comment (fruffy) [inline sim/src/node.rs:79]: What makes this necessary?

**Review by fruffy — CHANGES_REQUESTED** [review]

File: sim/src/main.rs (hunk 3 above)
Comment (jessberg) [reply sim/src/main.rs]: There is a sink that is connected to everything else.  Because it is everyone's neighbor, eventually the RPC will randomly go to the sink, where it disappears.

Comment (fruffy) [reply sim/src/main.rs]: I think sink should be an implicit thing. For example, in the [bookinfo](https://istio.io/latest/docs/examples/bookinfo/) app ratings and details should be sinks.

File: sim/src/main.rs (hunk 2 above)
Comment (jessberg) [reply sim/src/main.rs:78]: What do you mean by requests?  The rpcs have to be able to be sent to storage - that's how storage gets information.

File: sim/src/main.rs (hunk 3 above)
Comment (jessberg) [reply sim/src/main.rs]: Do you think we should make it an option when creating a node?

File: sim/src/node.rs (hunk 5 above)
Comment (jessberg) [reply sim/src/node.rs:79]: It's not - I previously had subsequent prints and forgot to revert this once I got rid of the prints.

File: sim/src/main.rs (hunk 2 above)
Comment (fruffy) [reply sim/src/main.rs:78]: The way storage works in the kubernetes env is an off-path request. Storage is not really part of the network (in the sense that requests by clients may travel there). Instead the proxies (and only the proxies) explicitly accesse storage with a custom rpc call.

File: sim/src/main.rs (hunk 3 above)
Comment (fruffy) [reply sim/src/main.rs]: Right now a sink just has zero output right? So an option does not seem necessary. Once we get to the point where nodes respond on requests we may have to rethink this.

File: sim/src/main.rs (hunk 2 above)
Comment (jessberg) [reply sim/src/main.rs:78]: Then I think requests can be sent to storage;  they have to be explicitly addressed to storage though.

File: sim/src/main.rs (hunk 1 above)
Comment (jessberg) [reply sim/src/main.rs:64]: I had seen some of the example queries use terms from hipster shop;  now that we have realized the example queries were messed up anyway, the choice has no reason behind it.  I changed it back.

File: sim/src/main.rs (hunk 3 above)
Comment (jessberg) [reply sim/src/main.rs]: So do you still think we should have an implicit sink?  There is no sense of the "end" of a graph because an RPC can be sent back to mimic a response, so we'd implicitly be adding it to all nodes in the simulator.

File: sim/src/main.rs (hunk 2 above)
Comment (fruffy) [reply sim/src/main.rs:78]: From the applications point of view storage may as well not exist. But the way this is modelled right now is that every regular node has a link to storage. This seems to apply that RPCs may be sent at random to to storage from these nodes, which is not what we want. This should be controlled by the filter.

File: sim/src/main.rs (hunk 3 above)
Comment (fruffy) [reply sim/src/main.rs]: Yes, until we have implemented responses these should be implicit sinks. That gets close to a request graph at least. We will have to develop custom node logic soon to model the full productpage application anyway.

File: sim/src/main.rs (hunk 2 above)
Comment (jessberg) [reply sim/src/main.rs:78]: Would it make more sense to allow edges between plugins and storage?  In that case, the simulator has to be aware of storage as a different entity, because it has to make those edges.  So either the simulator is aware of storage and its special status, or the user has to manually make edges between all the plugins and storage.  I don't like either of those designs but I'm not sure how else one could get a direct link between plugins and storage.

Another option is to have nodes be aware of what storage is, which is similarly untidy.

File: sim/src/main.rs (hunk 3 above)
Comment (jessberg) [reply sim/src/main.rs]: Okay, it is created implicitly now.

File: sim/src/main.rs (hunk 2 above)
Comment (fruffy) [reply sim/src/main.rs:78]: Is it possible to directly access storage from the filter? Logically speaking, this is just an API call after all. Alternatively, is it possible to directly queue requests in storage from a node?

Comment (fruffy) [reply sim/src/main.rs:78]:
```
    auto result = root()->httpCall("storage-upstream",
                                   { {":method", "GET"},
                                    {":path", "/store"},
                                    {":authority", "storage-upstream"},
                                    {"key", key},
                                    {"value", value} },
                                   "", {}, 1000, callback);
```
Looking at the filter this is basically a lookup request for the DNS server, which in our case would be a lookup in the map of simulator nodes.

File: sim/src/simulator.rs [hunk 6]
```diff
@@ -28,12 +29,14 @@ pub struct Simulator {
...
     pub fn new(seed: u64) -> Self {
-        Simulator {
+        let mut sim = Simulator {
             elements: HashMap::new(),
             graph: Graph::new(),
             node_index_to_node: HashMap::new(),
             seed,
-        }
+        };
+        sim.add_node(SINK_NAME, 1, 0, 0, None);
```
Comment (fruffy) [inline sim/src/simulator.rs]: Why is this needed? I thought you could just set the generation rate of ratings and details to 0?

File: sim/src/main.rs (hunk 2 above)
Comment (jessberg) [reply sim/src/main.rs:78]: Directly from the filter would be a little difficult, I think, because we'd be importing parts of the simulator into the filter.  That seems messy.

What do you mean by directly queue requests in storage from a node?   Right now, we are sending any RPCs from the filter with the destination storage to storage using the node-storage edge.  I don't see how that is any different.  We don't have a lookup in the map of simulator nodes, because we are only allowed to send to nodes that are neighbors.  Although the simulator is slightly more restrictive, I also think it makes sense because we have no application logic, so if we just chose a random node to send everything to, there would be no meaning to the structure of the graph.

Comment (fruffy) [reply sim/src/main.rs:78]: The graph has meaning to the nodes, but it does not necessarily have to have meaning for things like storage. My initial view was that plugin code could resolve names like "storage" from a central service, get the storage object, and directly query it. But it is probably better to do this over the queue paradigm. Is it possible for the plugin to queue an rpc with destination to storage, which is then picked up by the simulator, disregarding whether there is a connection or not? It should be, right? We just need to remove the check in the plugin wrapper tick. Otherwise I do not see how you can handle this without explicitly attaching storage to every plugin that is created.

Alternatively, you have to implement an explicit check in all nodes to **not** send to storage. Technically, that is also possible since we will implement custom logic for these nodes soon, anyway. Then there will be "random" nodes, which fire RPCs everywhere (including storage), and the bookinfo nodes, which have a specific request/response flow and ignore something like storage.

Comment (jessberg) [reply sim/src/main.rs:78]: I think it would be possible.  It would mean that we add a destination for every node when they send RPCs off, which isn't a bad idea anyway, if we want to eventually do routing.  

The easiest solution right now is to just add a check saying don't send to storage unless explicitly addressed in all nodes.  But I think the better solution would be to mandate that all RPCs have destinations for the next hop, and sources where they're from.  Nodes would automatically send all RPCs to their plugins upon receiving them, and wait for the next tick to get those RPCs back.  They can tell that an RPC has already been through the filter by the source node.  Then nodes will send to their neighbors.

Edges can pick from their neighbors for a destination, and plugins can choose between the node they're associated with and storage.  Then we can reintroduce plugins as their own elements in the simulator, and instead of making them a field of nodes, we can just make the plugin ID a field of a node, and that's a bit more modular anyway.  In addition, having a plugin will add an extra tick, and simulating some delay time associated with having a filter makes sense anyway.  

I think the only downside of this idea is that we'll have to hard code an ID for storage.  Does that seem like a reasonable solution?

File: sim/src/simulator.rs (hunk 6 above)
Comment (jessberg) [reply sim/src/simulator.rs]: Generation rate is how many RPCs are created out of thin air, not how many are passed on.  The egress rate is how many are passed on, but I don't think it's a good idea to set it to 0, because we're modeling a response being sent back as well.  So an RPC can optionally go back up the tree.

File: sim/src/main.rs (hunk 2 above)
Comment (fruffy) [reply sim/src/main.rs:78]: I have been thinking about the second option a little, my proposal is the following:
Node is actually a RandomNode, because that is what it does. It sends random requests to anything that is attached. We can keep it as is, and attach storage for now and do not worry about the RPCs it receives. So there are no changes necessary. 

Going forward, when we implement something like bookinfo, we will have to implement custom logic. This includes the destination and request/response patterns. But this logic should only be implemented for the specific node we want to simulate. For example, "productinfo-v1" will send a request to details **and** one of reviews at random. When it gets back a response from either of them, it will send to "loadgen". We can do this in the next pull request.  

Also I assumed that we already have destinations? At least that is what I see from the simulator tick code.

File: sim/src/simulator.rs (hunk 6 above)
Comment (fruffy) [reply sim/src/simulator.rs]: Based on my comments in the other thread I would say both values should be set to 0 in ratings and details and this extra sink should be removed. We do not have a model of request and response yet. That we will implement soon. Let's assume this just models the request DAG that ends there.

File: sim/src/main.rs (hunk 2 above)
Comment (jessberg) [reply sim/src/main.rs:78]: That sounds like a good idea.  So when we implement application logic, we just implement another simulator element, and we can have all the nodes, including RandomNode, ProductPage, etc, implement some sort of Node trait.

We have destinations as a "what is the next hop, please go there."  This gets you from a node to an edge, or edge to node.  We do not have routing as in "I am productpage and I want to go to node ratings-v1."  You can't go to a node you do not have a direct connection with, even if it is possible through multiple hops.  However, I think we only need the former, at least for now.

File: sim/src/simulator.rs (hunk 6 above)
Comment (jessberg) [reply sim/src/simulator.rs]: Okay, I changed it back.

File: sim/src/node.rs (hunk 5 above)
Comment (fruffy) [reply sim/src/node.rs:79]: There is one clone left?

**Review by fruffy — APPROVED** [review]

File: sim/src/node.rs (hunk 4 above)
Comment (jessberg) [reply sim/src/node.rs]: We have to either clone here or at ret.push as was below.  This is because we are using dest both for the print and moving it into ret.  I think it is better that we clone in the actual code than for a print, so I reverted the below change and got rid of the clone here.
